code/go/0chain.net/blobbercore/reference/referencepath.go

```diff
@@ -18,8 +18,71 @@ type ReferencePath struct {
 	Ref  *Ref
 }
 
-func GetReferencePath(ctx context.Context, allocationID, path string) (*Ref, error) {
-	return GetReferencePathFromPaths(ctx, allocationID, []string{path})
+//func GetReferencePath(ctx context.Context, allocationID, path string) (*Ref, error) {
+//	return GetReferencePathFromPaths(ctx, allocationID, []string{path})
+//}
+
+func GetReferencePath2(ctx context.Context, allocationID, path string) (*Ref, error) {
+	return GetReferenceForHashCalculationFromPaths(ctx, allocationID, []string{path})
+}
+
+// GetReferenceForHashCalculationFromPaths validate and build full dir tree from db, and CalculateHash and return root Ref without saving in DB
+func GetReferenceForHashCalculationFromPaths(ctx context.Context, allocationID string, paths []string) (*Ref, error) {
+	var refs []Ref
+	db := datastore.GetStore().GetTransaction(ctx)
+	db = db.Select("id", "allocation_id", "type", "name", "path",
+		"parent_path", "size", "hash", "path_hash", "content_hash", "merkle_root",
+		"actual_file_size", "actual_file_hash", "attributes", "chunk_size",
+		"lookup_hash", "thumbnail_hash")
+	pathsAdded := make(map[string]bool)
+	for _, path := range paths {
+		path = strings.TrimSuffix(path, "/")
+		if _, ok := pathsAdded[path]; !ok {
+			db = db.Where(Ref{ParentPath: path, AllocationID: allocationID})
+			pathsAdded[path] = true
+		}
+		depth := len(GetSubDirsFromPath(path)) + 1
+		curPath := filepath.Dir(path)
+		for i := 0; i < depth-1; i++ {
+			if _, ok := pathsAdded[curPath]; !ok {
+				db = db.Or(Ref{ParentPath: curPath, AllocationID: allocationID})
+				pathsAdded[curPath] = true
+			}
+			curPath = filepath.Dir(curPath)
+		}
+	}
+	// root reference_objects with parent_path=""
+	db = db.Or("parent_path = ? AND allocation_id = ?", "", allocationID)
+	err := db.Order("path, level").Find(&refs).Error
+	if err != nil {
+		return nil, err
+	}
+	// there is no any child reference_objects for affected path, and insert root reference_objects
+	if len(refs) == 0 {
+		return &Ref{Type: DIRECTORY, AllocationID: allocationID, Name: "/", Path: "/", ParentPath: "", PathLevel: 1}, nil
+	}
+
+	rootRef := &refs[0]
+	if rootRef.Path != "/" {
+		return nil, common.NewError("invalid_dir_tree", "DB has invalid tree. Root not found in DB")
+	}
+
+	// valdiate dir tree, and populate Ref's children for CalculateHash
+	refMap := make(map[string]*Ref)
+	refMap[rootRef.Path] = rootRef
+	for i := 1; i < len(refs); i++ {
+		if _, ok := refMap[refs[i].ParentPath]; !ok {
+			return nil, common.NewError("invalid_dir_tree", "DB has invalid tree.")
+		}
+		if _, ok := refMap[refs[i].Path]; !ok {
+			refMap[refs[i].ParentPath].AddChild(&refs[i])
+			refMap[refs[i].Path] = &refs[i]
+		}
+	}
+	if _, err := refs[0].CalculateHash(ctx, false); err != nil {
```

Comment:
if we are not storing it in db then do we even need to calculate hash? Its already there, right?